File: 2024/day-2/part_2.py
def is_safe(arr: list[int]) -> bool:
    is_increasing = is_decreasing = True

    for _ in range(len(arr) - 1):
        diff: int = arr[_] - arr[_ + 1]

        if not diff or diff < -3 or diff > 3:
            return False

        if diff < 0:
            is_decreasing = False
        else:
            is_increasing = False

        if not is_increasing and not is_decreasing:
            return False

    return True


def apply_dampener(arr: list[int]) -> bool:
    for _ in range(len(arr)):
        if is_safe(arr[:_] + arr[_ + 1:]):
            return True

    return False


def main() -> None:
    safe_count: int = 0

    with open("input", "r", encoding="UTF-8") as file:
        for line in file:
            array: list[int] = list(map(int, line.strip().split()))

            (is_safe(array) or apply_dampener(array)) and (safe_count := safe_count + 1)

    print(safe_count)

    return None


if __name__ == "__main__":
    main()

File: 2024/day-2/part_2.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

bool isSafe(int[], int);
bool applyDampener(int[], int);

int main(void) {
  FILE *file = fopen("input", "r");

  if (!file) {
    perror("File error");

    return EXIT_FAILURE;
  }

  char string[50];
  int array[25];
  int safeCount = 0;

  while (fgets(string, 50, file)) {
    int index = 0;
    char *token = strtok(string, " ");

    while (token) {
      array[index++] = atoi(token);
      token = strtok(NULL, " ");
    }

    (isSafe(array, index) || applyDampener(array, index)) && ++safeCount;
  }

  printf("%d\n", safeCount);
  fclose(file);

  return EXIT_SUCCESS;
}

bool isSafe(int arr[], int size) {
  bool isIncreasing = true, isDecreasing = true;

  for (int i = 0; i < size - 1; ++i) {
    int diff = arr[i] - arr[i + 1];

    if (!diff || diff < -3 || diff > 3)
      return false;

    if (diff < 0)
      isDecreasing = false;
    else
      isIncreasing = false;

    if (!isDecreasing && !isIncreasing)
      return false;
  }

  return true;
}

bool applyDampener(int arr[], int size) {
  int newArr[24];

  for (int i = 0; i < size; ++i) {
    int idx = 0;

    for (int j = 0; j < size; ++j)
      if (i != j)
        newArr[idx++] = arr[j];

    if (isSafe(newArr, idx))
      return true;
  }

  return false;
}